x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java

```diff
@@ -418,27 +476,114 @@ public void authenticate(final AuthenticationToken authenticationToken, final Ac
+                        now,
+                        header.getAlgorithm(),
+                        claimsSet.getIssuer(),
+                        claimsSet.getAudience(),
+                        header.getType(),
+                        claimsSet.getDateClaim("auth_time"),
+                        claimsSet.getIssueTime(),
+                        claimsSet.getNotBeforeTime(),
+                        claimsSet.getExpirationTime(),
+                        header.getKeyID(),
+                        claimsSet.getJWTID()
+                    );
+                }
+                // Validate all else before signature, because these checks are more helpful diagnostics than rejected signatures.
+                final boolean isJwtSigPkc = JwtRealmSettings.SUPPORTED_SIGNATURE_ALGORITHMS_PKC.contains(header.getAlgorithm().getName());
+                if (isJwtSigPkc && (this.isConfiguredJwkSetPkc == false)) {
+                    final String msg = "Realm [" + super.name() + "] not configured to verify PKC JWTs for token=[" + tokenPrincipal + "].";
+                    LOGGER.debug(msg);
+                    listener.onResponse(AuthenticationResult.unsuccessful(msg, null));
+                    return;
+                } else if ((isJwtSigPkc == false) && (this.isConfiguredJwkSetHmac == false) && (this.isConfiguredJwkOidcHmac == false)) {
+                    final String msg = "Realm ["
+                        + super.name()
+                        + "] not configured to verify HMAC JWTs for token=["
+                        + tokenPrincipal
+                        + "].";
+                    LOGGER.debug(msg);
+                    listener.onResponse(AuthenticationResult.unsuccessful(msg, null));
+                    return;
+                }
+
+                JwtValidateUtil.validateType(jwt);
+                JwtValidateUtil.validateIssuer(jwt, allowedIssuer);
+                JwtValidateUtil.validateAudiences(jwt, allowedAudiences);
+                JwtValidateUtil.validateSignatureAlgorithm(jwt, isJwtSigPkc ? this.allowedJwksAlgsPkc : this.allowedJwksAlgsHmac);
+                JwtValidateUtil.validateAuthTime(jwt, now, this.allowedClockSkew.seconds());
+                JwtValidateUtil.validateIssuedAtTime(jwt, now, this.allowedClockSkew.seconds());
+                JwtValidateUtil.validateNotBeforeTime(jwt, now, this.allowedClockSkew.seconds());
+                JwtValidateUtil.validateExpiredTime(jwt, now, this.allowedClockSkew.seconds());
+
+                // At this point, client authc and JWT kty+alg+iss+aud+time filters passed. Do sig last, in case JWK reload is expensive.
+                try {
+                    JwtValidateUtil.validateSignature(
+                        jwt,
+                        isJwtSigPkc ? this.contentAndJwksAlgsPkc.jwksAlgs.jwks : this.contentAndJwksAlgsHmac.jwksAlgs.jwks
+                    );
+                } catch (Exception originalValidateSignatureException) {
```

Comment:
Catching `Exception` is too liberal. Is it possible to narraow it down to specific error type for reloading?

Reply:
I started with separate exceptions, and I ran into corner cases where each exception inside validateSignature requires reload. I had to trigger reload for any exception in validationSignature to make the tests work.

- If in-memory JWKs fail to verify newer JWT signature, reload is required to get newer JWKs. This is directly covered in tests.
- If in-memory JWKs filtered by newer JWT kid header, reload is required to get newer JWKs with the newer kid attributes. This is indirectly covered in tests.
- If in-memory JWKs empty, realm loaded an empty JWKSet. If provider accidentally posted an empty JWKSet and fixed it, reload is required to recover. This is directly covered in tests.

Reply:
If no follow up comment, then I assume this is resolved.